src/dos/program_setver.cpp

```diff
@@ -0,0 +1,744 @@
+		if (!preprocess_name(params[0], permissive)) {
+			return;
+		}
+		// It shouldn't be a directory
+		if (!params[0].empty() && params[0].back() == '\\') {
+			WriteOut(MSG_Get("SHELL_EXPECTED_FILE_NOT_DIR"));
+			return;
+		}
+	}
+
+	// Detect illegal switch combinations
+	if ((has_arg_batch || has_arg_paged) && (has_arg_delete || has_arg_global)) {
+		WriteOut(MSG_Get("SHELL_ILLEGAL_SWITCH_COMBO"));
+		return;
+	}
+
+	// Handle cases with no file parameters
+	if (params.empty()) {
+		if (has_arg_delete && has_arg_global) {
+			CommandDeleteGlobalOnly(has_arg_quiet);
+		} else if (has_arg_delete) {
+			CommandDeleteAll(has_arg_quiet);
+		} else if (has_arg_global) {
+			WriteOut(MSG_Get("SHELL_MISSING_PARAMETER"));
+		} else {
+			CommandPrintAll(has_arg_batch, has_arg_paged);
+		}
+		return;
+	}
+
+	// From now on at least 1 parameter is guaranteed to exist
+
+	// Detect illegal switch combinations
+	if (has_arg_delete && has_arg_global) {
+		WriteOut(MSG_Get("SHELL_ILLEGAL_SWITCH_COMBO"));
+		return;
+	}
+
+	// Hande entry deletion
+	if (has_arg_delete) {
+		if (params.size() > 1) {
+			WriteOut(MSG_Get("SHELL_TOO_MANY_PARAMETERS"));
+		} else {
+			CommandDeletePerFile(params[0], has_arg_quiet);
+		}
+		return;
+	}
+
+	// Hande global version set
+	if (has_arg_global) {
+		if (params.size() > 1) {
+			WriteOut(MSG_Get("SHELL_TOO_MANY_PARAMETERS"));
+		} else {
+			CommandSet({}, params[0], has_arg_quiet);
+		}
+		return;
+	}
+
+	// Handle per-file version set
+	if (has_arg_batch || has_arg_paged) {
+		WriteOut(MSG_Get("SHELL_WRONG_SYNTAX"));
+	} else if (params.size() > 2) {
+		WriteOut(MSG_Get("SHELL_TOO_MANY_PARAMETERS"));
+	} else {
+		CommandSet(params[0], params[1], has_arg_quiet);
+	}
+}
+
+bool SETVER::ParseVersion(const std::string& version_str, FakeVersion& version)
+{
+	version.major = 0;
+	version.minor = 0;
+
+	const std::regex expression("^([0-9])(\\.([0-9]{0,2}))?$");
+	std::smatch match = {};
+
+	std::regex_match(version_str, match, expression);
+
+	if (match.size() != 4) {
+		return false;
+	}
+
+	const auto& major_str = match[1].str();
+	const auto& minor_str = match[3].str();
+
+	const auto major = to_int(major_str);
+	if (!major) {
+		// It would be enough to assert, but PVS-Studio was unhappy
+		assert(false);
+		return false;
+	}
+
+	version.major = static_cast<uint8_t>(*major);
+
+	if (minor_str.empty()) {
+		return true;
+	}
+
+	const auto minor = to_int(minor_str);
+	if (!minor) {
+		// It would be enough to assert, but PVS-Studio was unhappy
+		assert(false);
+		return false;
+	}
+
+	if (minor_str.size() == 1) {
+		version.minor = static_cast<uint8_t>(*minor * 10);
+	} else {
+		version.minor = static_cast<uint8_t>(*minor);
+	}
+
+	return true;
+}
+
+bool SETVER::PreprocessName(std::string& name, const bool permissive)
+{
+	// Preprocess file names and relative paths
+	trim(name);
+	if (IsNameWithPath(name)) {
+		char buffer[DOS_PATHLENGTH];
+		if (!DOS_Canonicalize(name.c_str(), buffer)) {
+			return permissive;
```

Comment:
Can "permissive" be named better? (Even in the function body, it doesn't permit or deny any code paths, and instead only controls the return values if we failed to canoicalize).

It seems like three return types are desired? fully processed, couldn't canonicalize, and failed/illegal. (maybe an enum return is better than the taking in the permissive bool?) -- just throwing out ideas if they can help.

Reply:
Changed variable name to `allow_non_existing_files`. Enum return requires a definition + more complicated result checking.